File: internal/crdt/vectors_test.go
package crdt

import (
	"encoding/json"
	"os"
	"path/filepath"
	"testing"
)

// testVectorFile mirrors the shared JSON test vector format documented
// in docs/crdt-specification.md ("Cross-Language Test Vectors") and
// consumed independently by both this package and web/src/lib/crdt.
// Operations are stored as raw JSON so each one can be fed directly to
// UnmarshalOperation -- the exact function real operations decode
// through, not a parallel test-only parser.
type testVectorFile struct {
	Description string            `json:"description"`
	DocumentID  string            `json:"documentId"`
	Operations  []json.RawMessage `json:"operations"`
	Expected    string            `json:"expected"`
}

func TestCrossLanguageVectors(t *testing.T) {
	dir := "../../testdata"
	entries, err := os.ReadDir(dir)
	if err != nil {
		t.Fatalf("reading testdata directory: %v", err)
	}

	var found int
	for _, entry := range entries {
		if entry.IsDir() || filepath.Ext(entry.Name()) != ".json" {
			continue
		}
		found++
		filename := entry.Name()

		t.Run(filename, func(t *testing.T) {
			data, err := os.ReadFile(filepath.Join(dir, filename))
			if err != nil {
				t.Fatalf("reading %s: %v", filename, err)
			}

			var vector testVectorFile
			if err := json.Unmarshal(data, &vector); err != nil {
				t.Fatalf("parsing %s: %v", filename, err)
			}

			doc := NewDocument(vector.DocumentID)
			for i, raw := range vector.Operations {
				op, err := UnmarshalOperation(raw)
				if err != nil {
					t.Fatalf("%s: decoding operation %d: %v", filename, i+1, err)
				}
				if err := doc.Apply(op); err != nil {
					t.Fatalf("%s: applying operation %d: %v", filename, i+1, err)
				}
			}

			if got := doc.Materialize(); got != vector.Expected {
				t.Errorf("%s: Materialize() = %q, want %q\n  description: %s",
					filename, got, vector.Expected, vector.Description)
			}
		})
	}

	if found == 0 {
		t.Fatal("no test vector files found in testdata -- did the directory move?")
	}
	t.Logf("verified %d cross-language test vectors", found)
}

File: web/src/lib/crdt/vectors.test.ts
import { describe, expect, it } from "vitest";
import { readdirSync, readFileSync } from "fs";
import path from "path";
import { Document } from "./document";
import { unmarshalOperation } from "./serialization";

// Loads and applies the same shared JSON test vectors that internal/crdt/vectors_test.go consumes independently on the Go side.

interface TestVectorFile {
  description: string;
  documentId: string;
  operations: unknown[];
  expected: string;
}

const testdataDir = path.resolve(import.meta.dirname, "../../../../testdata");

describe("cross-language test vectors", () => {
  const filenames = readdirSync(testdataDir).filter((f) => f.endsWith(".json"));

  it("found at least one vector file", () => {
    expect(filenames.length).toBeGreaterThan(0);
  });

  it.each(filenames)("%s", (filename) => {
    const raw = readFileSync(path.join(testdataDir, filename), "utf-8");
    const vector = JSON.parse(raw) as TestVectorFile;

    const doc = new Document(vector.documentId);
    vector.operations.forEach((rawOp, i) => {
      const op = unmarshalOperation(JSON.stringify(rawOp));
      try {
        doc.apply(op);
      } catch (err) {
        throw new Error(`${filename}: applying operation ${i + 1}: ${(err as Error).message}`);
      }
    });

    expect(doc.materialize(), `${filename}: ${vector.description}`).toBe(vector.expected);
  });
});